stdlib/public/Concurrency/Task.cpp

```diff
@@ -21,51 +21,150 @@
 #include "TaskPrivate.h"
 
 using namespace swift;
+using FutureFragment = AsyncTask::FutureFragment;
+
+void FutureFragment::destroy() {
+  auto queueHead = waitQueue.load(std::memory_order_acquire);
+  switch (queueHead.getStatus()) {
+  case Status::Executing:
+    assert(false && "destroying a task that never completed");
+
+  case Status::Success:
+    resultType->vw_destroy(getStoragePtr());
+    break;
+
+  case Status::Error:
+    swift_unknownObjectRelease(reinterpret_cast<OpaqueValue *>(getError()));
+    break;
+  }
+}
+
+FutureFragment::Status AsyncTask::waitFuture(AsyncTask *waitingTask) {
+  using Status = FutureFragment::Status;
+  using WaitQueueItem = FutureFragment::WaitQueueItem;
+
+  assert(isFuture());
+  auto fragment = futureFragment();
+
+  auto queueHead = fragment->waitQueue.load(std::memory_order_acquire);
+  while (true) {
+    switch (queueHead.getStatus()) {
+    case Status::Error:
+    case Status::Success:
+      // The task is done; we don't need to wait.
+      return queueHead.getStatus();
+
+    case Status::Executing:
+      // Task is now complete. We'll need to add ourselves to the queue.
+      break;
+    }
+
+    // Put the waiting task at the beginning of the wait queue.
+    waitingTask->getNextWaitingTask() = queueHead.getTask();
+    auto newQueueHead = WaitQueueItem::get(Status::Executing, waitingTask);
+    if (fragment->waitQueue.compare_exchange_weak(
+            queueHead, newQueueHead, std::memory_order_release,
+            std::memory_order_acquire)) {
+      // Escalate the priority of this task based on the priority
+      // of the waiting task.
+      swift_task_escalate(this, waitingTask->Flags.getPriority());
+      return FutureFragment::Status::Executing;
+    }
+  }
+}
+
+void AsyncTask::completeFuture(AsyncContext *context, ExecutorRef executor) {
+  using Status = FutureFragment::Status;
+  using WaitQueueItem = FutureFragment::WaitQueueItem;
+
+  assert(isFuture());
+  auto fragment = futureFragment();
+
+  // If an error was thrown, save it in the future fragment.
+  auto futureContext = static_cast<FutureAsyncContext *>(context);
+  bool hadErrorResult = false;
+  if (auto errorObject = futureContext->errorResult) {
+    fragment->getError() = errorObject;
+    hadErrorResult = true;
+  }
+
+  // Update the status to signal completion.
+  auto newQueueHead = WaitQueueItem::get(
+    hadErrorResult ? Status::Error : Status::Success,
+    nullptr
+  );
+  auto queueHead = fragment->waitQueue.exchange(
+      newQueueHead, std::memory_order_acquire);
+  assert(queueHead.getStatus() == Status::Executing);
+
+  // Schedule every waiting task on the executor.
+  auto waitingTask = queueHead.getTask();
+  while (waitingTask) {
+    // Find the next waiting task.
+    auto nextWaitingTask = waitingTask->getNextWaitingTask();
+
+    // Remove this task from the list.
+    waitingTask->getNextWaitingTask() = nullptr;
```

Comment:
You actually don't have to do this; the memory is invalidated once it's not waiting.

Reply:
Alright.